### ISSUE TITLE
Large Newgrounds Games not loading on iOS

### ISSUE THREAD

**Issue by PsychoGoldfishNG** [ruffle-rs/ruffle#6829]
### Describe the bug

When loading large games on Newgrounds from a 7th gen iPad (running iOS 15.4.1), one of three things happens:

1) Game actually works (rare)
2) Click on un-mute button appears, but does nothing when clicked
3) Gets stuck as a white screen (most consistent)

### Expected behavior

In case 2, 'Click to Unmute' shouldn't need to appear at all if audio is already initialized, but when it does, it should do something when clicked.

In case 3 The game should actually start after 'loadmetadata' is dispatched.

### Affected platform

Self-hosted version

### Operating system

iOS 15.4.1

### Browser

Google Chrome (Safari)

### Additional information

I have been using Chrome for iOS, which lets you view the console by running chrome://inspect in a new tab, however all browsers on iOS are just wrappers for Safari.

In case 2, I'm not getting anything in the console at all.

In case 3, the inspector in chrome is showing the following error, and it loops forever:
**ERROR** Unable to lock Ruffle core

I get no line numbers or anything helpful like that, but I can confirm it happens after the 'loadedmetadata' event is fired and readyState is set to 2.

I am hypothesizing that this may be caused by how Safari caches canvas elements, leading to a memory leak.
https://pqina.nl/blog/total-canvas-memory-use-exceeds-the-maximum-limit/

This behavior may also be related to the current preloading process we are using on Newgrounds, although I haven't been able to find any correlation that makes sense, and have had issues loading big games on NG prior to this update.  But just in case, here's an outline of what we're doing:

First we sideload the ruffle.js file (we do this to make sure it's loaded before the user tries to start a game) and render a custom play button with the game's icon.

When the end-user clicks this play button, we show our custom loader screen (it's just a <div> that gets displayed over the ruffle container), we preload the swf file over XHR (so we can show a progress bar), and we create, and play, an <audio> tag with a blank sound to pre-enable sound playback.  

Once everything is preloaded and initialized, we create and start the ruffle player and listen for the 'loadedmetadata' event to fire, at which point we remove the loading screen element from the DOM.

**Comment by danielhjacobs** [comment]
I believe this is actually an iOS bug: #6608

**Comment by PsychoGoldfishNG** [comment]
I was just thinking more about this.  As I mentioned in the ticket, I've had issues with large games not working even before we added our loading screen, but I did notice it was usually worse after I got it to work a few times.  And now that we are preloading the swf file before Ruffle starts, and it feels even more common, it got me thinking it may be having problems using large files from the disk cache on iOS.

Could also be something crapping out when the sound is starting.

**Comment by danielhjacobs** [comment]
Actually the bugs in the issues linked from that issue were marked as resolved but I don't know if those resolutions have shipped yet.

**Comment by Herschel** [comment]
Unfortunately iOS 15.4 has been completely busted for us, and we're patiently for 15.5 to release with fixes for the linked issues. I'll check some games on NG later tonight with my iPad to verify it's the same issue. It's possible there are other issues on top of this -- it's a little suspect to me that the game works sometimes, but maybe this is related to how aggressively the browser tries to JIT the WebAssembly, or something.

@PsychoGoldfishNG Do you know if any Unity WebGL games on NG are also affected? Ty for the report!

**Comment by PsychoGoldfishNG** [comment]
@Herschel Honestly, I gave up trying to play Unity games on my iPad long ago, so I couldn't say for sure lol.

**Comment by pestoforce** [comment]
I've been getting a panic error showing up on my iPad with the latest Ruffle nightly:

```
# Error Info
Error name: Error
Error message: panicked at 'Bezier without endpoint', core/src/shape_utils.rs:213:36
Error stack:
```
@https://typehelp.org/ruffletest/ruffle/core.ruffle.35ba130500fa865e8e67.js:1:7717
wasm-stub@[wasm code]
<?>.wasm-function[std::sync::once::Once::call_once::{{closure}}::h802bfeb6f8bb5638]@[wasm code]
<?>.wasm-function[std::sync::once::Once::call_inner::h07d0245bef259c27]@[wasm code]
<?>.wasm-function[core::ops::function::FnOnce::call_once{{vtable.shim}}::h9ecdeaaea35085f8]@[wasm code]
<?>.wasm-function[std::panicking::rust_panic_with_hook::h623ac32ff431b114]@[wasm code]
<?>.wasm-function[core::panicking::panic_fmt::h5118e89563022e7e]@[wasm code]
<?>.wasm-function[core::option::expect_failed::h52436cb58dfe4ec3]@[wasm code]
<?>.wasm-function[<core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next::hed214b29a827cdd6]@[wasm code]
<?>.wasm-function[ruffle_core::shape_utils::ShapeConverter::flush_layer::h66beb67534a7dd6e]@[wasm code]
<?>.wasm-function[<ruffle_core::shape_utils::DistilledShape as core::convert::From<&swf::types::Shape>>::from::h6e92fa6608141852]@[wasm code]
<?>.wasm-function[ruffle_core::display_object::movie_clip::MovieClipData::define_shape::ha98dc4b982c17625]@[wasm code]
<?>.wasm-function[ruffle_core::display_object::movie_clip::MovieClip::preload::hbaf6f019c2eaff27]@[wasm code]
<?>.wasm-function[ruffle_core::player::Player::set_root_movie::h1a20254dba634c51]@[wasm code]
<?>.wasm-function[<core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h903ee168fc90b9ee]@[wasm code]
<?>.wasm-function[<core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h33ed72626bb34236]@[wasm code]
<?>.wasm-function[wasm_bindgen_futures::queue::Queue::new::{{closure}}::he8e3ce0ab6352214]@[wasm code]
<?>.wasm-function[<dyn core::ops::function::FnMut<(A,)>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::h40f6ecbecfe0d2cc]@[wasm code]
wasm-stub@[native code]
O@https://typehelp.org/ruffletest/ruffle/core.ruffle.35ba130500fa865e8e67.js:1:3868
c@https://typehelp.org/ruffletest/ruffle/core.ruffle.35ba130500fa865e8e67.js:1:2393
promiseReactionJob@[native code]
```

# Player Info
Allows script access: true
Player type: Object
SWF URL: pdtd.swf
Param movie: pdtd.swf
Param allowscriptaccess: always
Attribute 0: undefined
Attribute 1: undefined
Attribute 2: undefined
Attribute 3: undefined

# Page Info
Page URL: https://typehelp.org/ruffletest/pdtd.html
SWF URL: https://typehelp.org/ruffletest/pdtd.swf

# Browser Info
Useragent: Mozilla/5.0 (iPad; CPU OS 15_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/100.0.4896.85 Mobile/15E148 Safari/604.1
OS: iPad

# Ruffle Info
Version: 0.1.0
Name: nightly 2022-04-14
Channel: nightly
Built: 2022-04-14T00:15:14.709Z
Commit: 036aa2b369aa9105800fbd49954bd2ae42281a0a
Is extension: false
```

**Comment by Herschel** [comment]
Yea, I'm pretty sure that's the same Apple bug. I'll try to install the iOS 15.5 beta later tonight to verify that it'll be fixed in the next iOS update.

**Comment by Herschel** [comment]
Jus tried Portal Defenders some on my 2021 iPad w/ 15.4 and it's definitely the same Apple Wasm issue. In my particular case I get "out of bounds memory access", but the specific error probably varies based on hardware.

I also tried a Unity WebGL game (https://www.newgrounds.com/portal/view/841329) on my iPad OS 15.4, and it crashed in a similar way, so not only a Ruffle issue.

Unfortunately I then installed iPad OS 15.5 beta 3 and still had the same issues with both games. I sure hope the fix ships in public 15.5, which is supposed to be sometime in May.

Going to close this in favor of #6608.
